docs/developers/design_docs/execution_model.md

```diff
@@ -0,0 +1,478 @@
+# IREE Invocation Execution Model
+
+This documents the behavior of the user-visible invocation mechanism IREE uses
+to schedule program execution. Internally IREE uses a very similar modeling for
+tracking its internal workloads and in kind carries that down to target APIs
+and devices that themselves use a very similar model. The intent is to expose
+the device model in an abstracted way that allows for the full capture and
+communication of the execution intent to be propagated to the hardware that
+executes it. Though here we focus on the user-visible portion of execution
+there is really only one "IREE execution model" and the entire stack follows
+the same design. At its core this design is just an instantiation of an
+[out-of-order execution](https://en.wikipedia.org/wiki/Out-of-order_execution) algorithm such as those [originating from the 1960's](https://en.wikipedia.org/wiki/Tomasulo_algorithm).
+
+## Glossary
+
+```mermaid
+stateDiagram
+    state UserApplication {
+      direction BT
+      state Context0 {
+        ModuleA-->ModuleAState0
+        ModuleB-->ModuleBState0
+      }
+      state Context1 {
+        ModuleA-->ModuleAState1
+        ModuleB-->ModuleBState1
+        ModuleC-->ModuleCState1
+      }
+      state ModuleA {
+        @func1
+        @func2
+      }
+      state ModuleB {
+        @func3
+        @func4
+      }
+      state ModuleC {
+        @func5
+      }
+    }
+```
+
+### Program
+
+An IREE _program_ is a collection of _modules_ instantiated in a _context_ from
+which _invocations_ can be made. Invocations are ordered on a user-controlled
+_timeline_ that uses _fences_ to define the execution order requirements to
+enable out-of-order execution. A hosting user application may have multiple
+programs or multiple instances of the same program available and running
+invocations at a time across multiple timelines.
```

Comment:
I'd like to refine https://iree-org.github.io/iree/bindings/c-api/ eventually, and there's some good language in here. We could start by linking to this new page ("for more details, see ...")

Reply:
For sure - I'm hoping to push through the API changes for stream-ordered allocations and fences and then even if the implementation isn't doing anything special it should let us put together docs/bindings using it 🤞